### PR TITLE
volume persistence database

### PR DESCRIPTION
**Pull request by heavenlyhash**
This adds use of boltdb to the volume management system.  Volume lifecycle events and metadata will be persisted so that volumes can be recalled and reconnected to jobs across flynn-host daemon reboots.

Also, a `flynn-host destroy-volumes` command has been added to remove state from previous hosts if a clean slate is desired.

### CODE REVIEW

File: host/volume/zfs/persistence_test.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+var _ = Suite(&PersistenceTests{})
+
+func (PersistenceTests) SetUpSuite(c *C) {
+	// Skip all tests in this suite if not running as root.
+	// Many zfs operations require root priviledges.
+	testutils.SkipIfNotRoot(c)
+}
+
+func (s *PersistenceTests) TestApplePie(c *C) {
```
Comment (titanous) [inline host/volume/zfs/persistence_test.go]: What does this test again?

File: host/volume/zfs/zfs.go [hunk 2]
```diff
@@ -110,77 +115,108 @@ func NewProvider(config *ProviderConfig) (volume.Provider, error) {
...
-		Location: path,
+func (b *Provider) MarshalGlobalState() (json.RawMessage, error) {
+	return json.Marshal(b.config)
+}
+
+func (b *Provider) MarshalVolumeState(volumeID string) (json.RawMessage, error) {
+	vol := b.volumes[volumeID]
+	c := make(map[string]string)
+	c["basemount"] = vol.basemount
+	c["dataset"] = vol.dataset.Name
```
Comment (titanous) [inline host/volume/zfs/zfs.go]: Probably should make a struct for this.

File: host/volume/zfs/persistence_test.go (hunk 1 above)
Comment (heavenlyhash) [inline host/volume/zfs/persistence_test.go]: Creating the universe from scratch :)

Kidding, placeholder needs rename badly.

**Comment by heavenlyhash** [comment]
Comments addressed and WIP commit replaced with actual description.

I think this is broadly ready for merge, with one major exception: currently, retaining etcd's state is causing the router application to crash on start.  As a result, though CI passes, running the integration tests twice in a row on the dev vm results in bootstrap failure.  I think this is the same root issue as #709 and will be fixed by #959.

**Comment by titanous** [comment]
Okay, can you add a quick command to flynn-host that nukes all of the volumes/state?

File: host/volume/manager/persistence_test.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+	volProv, err := zfs.NewProvider(&zfs.ProviderConfig{
+		DatasetName: zfsDatasetName,
+		Make: &zfs.MakeDev{
+			BackingFilename: zfsVdevFilePath,
+			Size:            int64(math.Pow(2, float64(30))),
+		},
+	})
+	c.Assert(err, IsNil)
+
+	// new volume manager with that shiney new backing zfs vdev file and a new boltdb
```
Comment (titanous) [inline host/volume/manager/persistence_test.go]: s/shiney/shiny/

**Comment by archSeer** [comment]
Very nice! Still reviewing, but it seems fine. I think we do need to address the issue of cleaning up behind the app, namely resources and routes.

**Comment by titanous** [comment]
LGTM

**Pull request merged**